### PR TITLE
Added Cancel

### PR DESCRIPTION
**Pull request by frecco**

### CODE REVIEW

File: Cancel/Creatable/index.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+
+export interface Creatable {
+	descriptor?: string
+}
+
+export namespace Creatable {
+	export function is(value: Creatable | any): value is Creatable {
+
```
Comment (simonmika) [inline Cancel/Creatable/index.ts]: Only return in one place. Use variable name `result` for it or in this case write it as a single expression.

File: Cancel/Creatable/index.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+		if ("descriptor" in value && typeof(value.descriptor) != "string") {
```
Comment (simonmika) [inline Cancel/Creatable/index.ts]: Write like this instead:
```
return typeof(value) == "object" &&
  (value.descriptor == undefined || typeof(value.descriptor) == "string") &&
  Object.keys(value).every(key => key == "descriptor")

File: Cancel/index.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import * as isoly from "isoly"
+import * as authly from "authly"
+import { Creatable as CancelCreatable } from "./Creatable"
+
+export interface Cancel {
+	id: authly.Identifier,
+	descriptor?: string,
+	created?: isoly.DateTime,
```
Comment (simonmika) [inline Cancel/index.ts]: `created` should be mandatory.

File: Cancel/index.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+export namespace Cancel {
+
+	export function is(value: Cancel | any): value is Cancel {
+		return true
```
Comment (simonmika) [inline Cancel/index.ts]: Implement.

File: Cancel/index.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	}
+
+	export function from(value: CancelCreatable): Omit<Cancel, "id"> {
+
```
Comment (simonmika) [inline Cancel/index.ts]: Superfluous space.

File: Cancel/index.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import { Creatable as CancelCreatable } from "./Creatable"
+
+export interface Cancel {
+	id: authly.Identifier,
+	descriptor?: string,
+	created?: isoly.DateTime,
+}
+
+export namespace Cancel {
+
```
Comment (simonmika) [inline Cancel/index.ts]: Superfluous space.

File: Cancel/index.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	export function is(value: Cancel | any): value is Cancel {
+		return true
+	}
+
+	export function from(value: CancelCreatable): Omit<Cancel, "id"> {
+
+		const capture = {
+			descriptor: value.descriptor,
+		}
+
```
Comment (simonmika) [inline Cancel/index.ts]: Superfluous space.

**Comment by simonmika** [comment]
Don't forget to add a `cancel` property to `Authorization` as well.

File: Authorization/index.ts [hunk 8]
```diff
@@ -15,6 +16,7 @@ export interface Authorization {
 	currency?: isoly.Currency
 	card: Card
 	capture: Capture[],
+	cancel?: Cancel,
```
Comment (simonmika) [inline Authorization/index.ts]: Can it not be more then one cancel per authorization? Cancelling some of the total amount.

File: Cancel/Creatable/index.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+
+export interface Creatable {
+	descriptor?: string
+}
+
+export namespace Creatable {
+	export function is(value: Creatable | any): value is Creatable {
+		let result = true
```
Comment (simonmika) [inline Cancel/Creatable/index.ts]: Write as a single expression.

File: index.ts [hunk 10]
```diff
@@ -1,14 +1,14 @@
 import * as Acquirer from "./Acquirer"
 import { Authorization } from "./Authorization"
-import { Capture } from "./Capture"
```
Comment (simonmika) [inline index.ts]: Don't remove Capture.

File: Authorization/index.ts (hunk 8 above)
Comment (frecco) [reply Authorization/index.ts]: No there cant be it seems.

Comment (simonmika) [reply Authorization/index.ts]: OK, let's figure out how to handle failed cancels in the future.

File: Cancel/index.ts (hunk 4 above)
Comment (simonmika) [reply Cancel/index.ts]: Remains

File: index.ts (hunk 10 above)
Comment (simonmika) [reply index.ts]: Remains

File: Cancel/index.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+	reference: string,
+	created: isoly.DateTime,
+}
+
+export namespace Cancel {
+	export function is(value: Cancel | any): value is Cancel {
+		return typeof value == "object" &&
+		       authly.Identifier.is(value.id) &&
+		       (value.descriptor == undefined || typeof value.descriptor == "string") &&
+		       typeof value.reference == "string" &&
```
Comment (simonmika) [inline Cancel/index.ts]: We have been using parenthesis around the argument to `typeof` but you are right, it is an operator and we should use it as one without the parenthesis.

**Pull request merged**